File: python/src/threads/reduce/transcript.py
"""ReducedState.transcript: the conversational entries the next Render v1 request renders.

Only which events render, and in what order, matters here, so no artifact is read. The render
rules are spec/schema/README.md, "Render v1".
"""

from collections.abc import Mapping, Sequence
from dataclasses import dataclass
from typing import Literal

from threads.log import (
    CompactedEvent,
    Event,
    EventId,
    HeartbeatEvent,
    InjectedEvent,
    MessageReceivedEvent,
    ModelResponseEvent,
    ModelResponseRecoveredEvent,
    SteerEvent,
    ThreadStartedEvent,
    ToolResultEvent,
    ToolResultLateEvent,
    UserInputEvent,
)
from threads.reduce.view import RenderView, assistant_parts, render_view, walk

type Role = Literal["user", "context", "assistant", "tool", "summary"]


@dataclass(frozen=True, slots=True)
class TranscriptEntry:
    role: Role
    event_id: EventId


# Yielded by the walk only in place of its range, a CompactedEvent is the summary.
_ROLES: Mapping[type, Role] = {
    InjectedEvent: "context",
    HeartbeatEvent: "context",
    MessageReceivedEvent: "context",
    ToolResultEvent: "tool",
    ToolResultLateEvent: "tool",
    CompactedEvent: "summary",
}


def _role(view: RenderView, event: Event) -> Role | None:
    if isinstance(event, UserInputEvent | SteerEvent):
        return None if event.event_id in view.denied else "user"
    if isinstance(event, ModelResponseEvent | ModelResponseRecoveredEvent):
        return "assistant" if assistant_parts(view, event) else None
    return _ROLES.get(type(event))


def transcript(events: Sequence[Event]) -> tuple[TranscriptEntry, ...]:
    if not any(isinstance(e, ThreadStartedEvent) for e in events):
        return ()  # a team log never renders
    view = render_view(events)
    entries: list[TranscriptEntry] = []
    for event in walk(view, events):
        role = _role(view, event)
        if role is not None:
            entries.append(TranscriptEntry(role, event.event_id))
    return tuple(entries)

File: typescript/packages/core/src/reduce/transcript.ts
import { assertNever } from "../assert-never";
import type { KnownEvent } from "../log";
import {
  assistantParts,
  type View,
  type VisibleEvent,
  view,
  walk,
} from "../render/view";

export type TranscriptEntry = {
  readonly role: "user" | "context" | "assistant" | "tool" | "summary";
  readonly event_id: string;
};

/**
 * ReducedState.transcript: `{role, event_id}` for each conversational line the next request
 * renders, in render order (spec/conformance/README.md; reference: render.py `transcript`).
 */
export function transcript(
  events: readonly KnownEvent[],
): readonly TranscriptEntry[] {
  // A team log has no thread_started: it never renders.
  if (!events.some((e) => e.type === "thread_started")) return [];
  const v = view(events);
  const out: TranscriptEntry[] = [];
  for (const entry of walk(v, events)) {
    if (entry.kind === "summary") {
      out.push({ role: "summary", event_id: entry.compacted.event_id });
      continue;
    }
    const role = roleOf(v, entry.event);
    if (role !== undefined) out.push({ role, event_id: entry.event.event_id });
  }
  return out;
}

function roleOf(v: View, e: VisibleEvent): TranscriptEntry["role"] | undefined {
  switch (e.type) {
    case "user_input":
    case "steer":
      return v.denied.has(e.event_id) ? undefined : "user";
    case "injected":
    case "heartbeat":
    case "message_received":
      return "context";
    case "model_response":
    case "model_response_recovered":
      return assistantParts(v, e).length > 0 ? "assistant" : undefined;
    case "tool_result":
    case "tool_result_late":
      return "tool";
    case "tools_changed":
      return undefined;
    default:
      return assertNever(e);
  }
}
